### PR TITLE
Refactor Data Input/Output Operators

### PR DESCRIPTION
**Pull request by jortiz16**
I renamed some operators as we discussed on Friday. I know this is still tentative, but it can be easily changed if we vote against it on Monday:

```DataInput``` --> ```TupleSource```
```DataOutput``` --> ```TupleSink```
```TupleSource``` --> ```BatchTupleSource``` (this class is usually used for testing, I wasn't sure what to call it but I was forced to rename it so it would not conflict with the new TupleSource class above)
```SinkRoot``` --> ```EmptySink```

This PR creates the new TupleSource operator, which eliminates the need for the FileScan and BinaryScan. Corresponding TupleReaders are created for these classes (CsvTupleReader and BinaryTupleReader). The other scans left are the Tipsy, NChilada and SeaFlow scans.

### CODE REVIEW

File: src/edu/washington/escience/myria/TupleReader.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+@JsonSubTypes({ @Type(name = "Csv", value = CsvTupleReader.class) })
+public interface TupleReader extends Serializable {
+
+  void open(InputStream stream) throws IOException, DbException;
+
+  Schema getSchema();
+
+  TupleBatch readTuples() throws IOException, DbException;
+
+  void done() throws IOException;
```
Comment (senderista) [inline src/edu/washington/escience/myria/TupleReader.java]: Any reason not to call this `close()` for symmetry with `open()`?

File: src/edu/washington/escience/myria/api/encoding/QueryConstruct.java [hunk 2]
```diff
@@ -313,7 +313,7 @@ private static void verifyAndPropagateLocalEdgeConstraints(final List<PlanFragme
 
   /**
    * Use the Catalog to set the workers for fragments that have scans, and verify that the workers are consistent with
-   * existing constraints.
+   * existing constraints. f
```
Comment (senderista) [inline src/edu/washington/escience/myria/api/encoding/QueryConstruct.java]: typo

File: src/edu/washington/escience/myria/operator/DataInput.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  /**
+   * Instantiate a new DataInput operator, which will stream its tuples to the specified {@link TupleReader}.
+   */
+  public DataInput(final TupleReader tupleReader, final DataSource dataSource) {
+    this.tupleReader = tupleReader;
+    this.dataSource = dataSource;
+  }
+
+  @Override
+  protected void init(final ImmutableMap<String, Object> execEnvVars) throws DbException {
```
Comment (senderista) [inline src/edu/washington/escience/myria/operator/DataInput.java]: Don't we also need to `close()` `tupleReader` in a `cleanup()` override?

File: systemtest/edu/washington/escience/myria/systemtest/JsonOperatorTests.java [hunk 4]
```diff
@@ -122,6 +123,6 @@ public void splitTest() throws Exception {
             .getProgramName(), outputRelation.getRelationName(), "json");
     String expectedData =
         "[{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"a\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"b\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"c\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"d\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"e\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"f\"}]";
-    assertEquals(expectedData, data);
+    assertEquals(data, expectedData);
```
Comment (senderista) [inline systemtest/edu/washington/escience/myria/systemtest/JsonOperatorTests.java]: The original order of args is correct: http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertEquals(java.lang.Object, java.lang.Object)

File: src/edu/washington/escience/myria/BinaryTupleReader.java [hunk 5]
```diff
@@ -41,32 +32,48 @@
   private transient DataInput dataInput;
 
   /**
+   * Construct a new BinaryFileScan object that reads the given binary file and creates tuples from the file data that
```
Comment (senderista) [inline src/edu/washington/escience/myria/BinaryTupleReader.java]: `BinaryTupleReader`

File: src/edu/washington/escience/myria/BinaryTupleReader.java [hunk 6]
```diff
@@ -41,32 +32,48 @@
...
-   * @param schema The tuple schema to be used for creating tuple from the binary file's data.
-   * @param source The source of the binary input data.
-   */
-  public BinaryFileScan(final Schema schema, final DataSource source) {
-    this(schema, source, false);
+  @Override
+  public void open(final InputStream stream) throws IOException, DbException {
+    buffer = new TupleBatchBuffer(getSchema());
+    InputStream inputStream;
+    inputStream = new BufferedInputStream(stream);
```
Comment (senderista) [inline src/edu/washington/escience/myria/BinaryTupleReader.java]: collapse into one line

File: src/edu/washington/escience/myria/BinaryTupleReader.java [hunk 7]
```diff
@@ -41,32 +32,48 @@
...
+   * has the given schema. The default endianess is big endian.
+   * 
+   * @param schema The tuple schema to be used for creating tuple from the binary file's data.
+   */
+  public BinaryTupleReader(final Schema schema) {
+    this(schema, false);
+  }
+
+  /**
    * Construct a new BinaryFileScan object that reads the given binary file and create tuples from the file data that
```
Comment (senderista) [inline src/edu/washington/escience/myria/BinaryTupleReader.java]: `BinaryTupleReader`

**Comment by senderista** [comment]
LGTM, we already reached consensus on renaming so I think this is ready to merge.

**Comment by jortiz16** [comment]
I also ended up running a rebase + force push for this PR. The only new relevant commit is the "fixing comments based on code review". I fixed the "BinaryFileScan" comments and I also fixed the first typo you described but I fixed it in an earlier commit (you won't see this change in the newest commit).

**Comment by jortiz16** [comment]
oops I missed other comments you had earlier... fixing now...

**Comment by jortiz16** [comment]
ok now it should be good

**Comment by senderista** [comment]
I forgot that the corresponding work in raco hasn't been done yet, so users already hit a regression. Need to revert this merge until the raco work is complete.

**Comment by senderista** [comment]
Since there's nothing wrong (AFAIK) with these changes, I'll just tag the branch before the merge and use that tag in the Myria deployment script until the raco changes are ready.

**Comment by jortiz16** [comment]
I completely forgot this as well! Thanks for catching this.

**Pull request merged**